File: otask-bot/bot/reminders.py
"""Фоновый планировщик напоминаний о дедлайнах."""
from __future__ import annotations

import logging
from datetime import datetime, timezone

from aiogram import Bot
from aiogram.enums import ParseMode

from .config import config
from .formatting import format_task_card
from .otask_client import OtaskClient, OtaskError
from . import store

logger = logging.getLogger("otask-bot.reminders")
_client = OtaskClient()


async def run_reminders_once(bot: Bot) -> None:
    """Один прогон: разослать напоминания всем подписанным чатам."""
    chats = await store.chats_for_reminders()
    if not chats:
        return

    now = datetime.now(timezone.utc)
    lead = config.reminder_lead_hours * 3600

    try:
        tasks = await _client.list_assigned_tasks()
    except OtaskError as e:
        logger.warning("Не удалось получить задачи: %s", e)
        return

    for task in tasks:
        if task.done or task.deadline is None:
            continue
        delta = (task.deadline - now).total_seconds()

        if delta < 0:
            await _notify_all(bot, chats, task, "overdue",
                              f"⚠️ <b>Задача просрочена!</b>\n\n{format_task_card(task, now)}")
        elif delta <= lead:
            left = max(1, round(delta / 3600))
            await _notify_all(bot, chats, task, "deadline",
                              f"⏰ <b>Скоро дедлайн</b> (через ~{left} ч)\n\n{format_task_card(task, now)}")
        else:
            # Дедлайн снова далеко (перенесли) — разрешим напомнить заново позже.
            for chat_id in chats:
                await store.clear_reminded(chat_id, task.id, "deadline")
                await store.clear_reminded(chat_id, task.id, "overdue")


async def _notify_all(bot: Bot, chats: list[int], task, kind: str, text: str) -> None:
    for chat_id in chats:
        if await store.mark_reminded(chat_id, task.id, kind):
            try:
                await bot.send_message(
                    chat_id, text, parse_mode=ParseMode.HTML, disable_web_page_preview=True
                )
            except Exception as e:  # noqa: BLE001
                logger.warning("Не смог отправить в чат %s: %s", chat_id, e)

File: otask-bot/src/reminders.ts
import type { Bot } from 'grammy';
import { config } from './config.js';
import { OtaskClient } from './otask/client.js';
import type { Task } from './otask/types.js';
import { store } from './store.js';
import { formatTaskCard } from './format.js';

/**
 * Периодически опрашивает otask и шлёт напоминания:
 *  • о задачах, у которых дедлайн наступит в ближайшие REMINDER_LEAD_HOURS;
 *  • о задачах, которые только что стали просроченными.
 * Дедупликация — через store, чтобы не спамить одинаковыми напоминаниями.
 */
export function startReminders(bot: Bot): () => void {
  const intervalMs = Math.max(1, config.reminderPollMinutes) * 60_000;
  let running = false;

  const tick = async () => {
    if (running) return; // не наслаиваем прогоны
    running = true;
    try {
      await runOnce(bot);
    } catch (e) {
      console.error('[reminders] прогон упал:', (e as Error).message);
    } finally {
      running = false;
    }
  };

  // Первый прогон — с небольшой задержкой после старта.
  const startTimer = setTimeout(tick, 10_000);
  const timer = setInterval(tick, intervalMs);
  return () => {
    clearTimeout(startTimer);
    clearInterval(timer);
  };
}

async function runOnce(bot: Bot): Promise<void> {
  const chats = await store.chatsForReminders();
  const now = new Date();
  const leadMs = config.reminderLeadHours * 3600_000;

  for (const chatId of chats) {
    const key = await store.getApiKey(chatId);
    if (!key) continue;

    let tasks: Task[];
    try {
      tasks = await new OtaskClient(key).listAssignedTasks();
    } catch (e) {
      console.error(`[reminders] chat ${chatId}: ${(e as Error).message}`);
      continue;
    }

    for (const task of tasks) {
      if (task.done || !task.deadline) continue;
      const dueMs = new Date(task.deadline).getTime();
      const delta = dueMs - now.getTime();

      if (delta < 0) {
        // Просрочена. Сбрасываем «предупреждали о дедлайне», шлём про просрочку один раз.
        if (await store.markReminded(chatId, task.id, 'overdue')) {
          await safeSend(bot, chatId, `⚠️ <b>Задача просрочена!</b>\n\n${formatTaskCard(task, now)}`);
        }
      } else if (delta <= leadMs) {
        // Скоро дедлайн.
        if (await store.markReminded(chatId, task.id, 'deadline')) {
          const left = Math.max(1, Math.round(delta / 3600_000));
          await safeSend(
            bot,
            chatId,
            `⏰ <b>Скоро дедлайн</b> (через ~${left} ч)\n\n${formatTaskCard(task, now)}`,
          );
        }
      } else {
        // Дедлайн снова далеко (его перенесли) — разрешим напомнить заново позже.
        await store.clearReminded(chatId, task.id, 'deadline');
        await store.clearReminded(chatId, task.id, 'overdue');
      }
    }
  }
}

async function safeSend(bot: Bot, chatId: number, text: string): Promise<void> {
  try {
    await bot.api.sendMessage(chatId, text, {
      parse_mode: 'HTML',
      link_preview_options: { is_disabled: true },
    });
  } catch (e) {
    console.error(`[reminders] не смог отправить в chat ${chatId}: ${(e as Error).message}`);
  }
}
